GVFS/GVFS.Virtualization/Projection/FileTypeAndMode.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+namespace GVFS.Virtualization.Projection
+{
+    public struct FileTypeAndMode
+    {
+        public static readonly ushort FileMode755 = Convert.ToUInt16("755", 8);
+        public static readonly ushort FileMode664 = Convert.ToUInt16("664", 8);
+        public static readonly ushort FileMode644 = Convert.ToUInt16("644", 8);
+
+        // Bitmasks for extracting file type and mode from the ushort stored in the index
+        private const ushort FileTypeMask = 0xF000;
+        private const ushort FileModeMask = 0x1FF;
+
+        // Values used in the index file to indicate the type of the file
+        private const ushort RegularFileIndexEntry = 0x8000;
+        private const ushort SymLinkFileIndexEntry = 0xA000;
+        private const ushort GitLinkFileIndexEntry = 0xE000;
+
+        public FileTypeAndMode(ushort typeAndModeInIndexFormat)
+        {
+            switch (typeAndModeInIndexFormat & FileTypeMask)
+            {
+                case RegularFileIndexEntry:
+                    this.Type = FileType.Regular;
+                    break;
+                case SymLinkFileIndexEntry:
+                    this.Type = FileType.SymLink;
+                    break;
+                case GitLinkFileIndexEntry:
```

Comment:
We don't actually support gitlinks right? If we ever see a repo with this data, we may want to fail hard as early as possible, rather than wait and see...

Reply:
I thought about making a change so that we'd throw when parsing the index if we encounter a git link (i.e. submodule).  

However, doing so would allow someone to make their repo unmountable:

1. Clone repo
2. Add git submodule
3. The next time VFS for Git parses the index it would throw an exception that results in GVFS.Mount.exe shutting down
4. Repo cannot be mounted or repaired

This also seemed outside the scope of adding symlink support, and so if we do want to add this validation (and prevent a repo from becoming unmountable) it's probably a big enough task to put in its own PR.